doc: Add launch script document (#7802)
projectacrn/acrn-hypervisor: doc/user-guides/acrn-dm-parameters.rst

.. _acrn-dm_parameters-and-launch-script:

Device Model Parameters and Launch Script
#########################################

Hypervisor Device Model (DM) is a QEMU-like application in the Service
VM responsible for creating a User VM and then performing devices
emulation based on command line configurations, as introduced in

> :ref:`hld-devicemodel`.

Reviewer: In this introduction, please describe why launch script is also described in this document.

Author: ok

   * - ``wdt-i6300esb``
     - Emulated i6300ESB PCI Watch Dog Timer (WDT) Intel processors use to
       monitor User VMs.

Launch Script
*************

Launch script is used to launch a User VM from the ServiceVM command line. It
is generated by configurator according to several settings for a User VM. With
launch script, user does not bother tweaking acrn-dm parameters.

Most settings for User VMs are used at launch time. They are first stored into
``scenario.xml`` and then parsed by
``misc/config_tools/launch_config/launch_cfg_gen.py``
to add shell commands into launch script according to the template
``misc/config_tools/launch_config/launch_script_template.sh``.
The template defines following helper functions to do system settings or to
generate an acrn-dm command option. For details about all acrn-dm parameters,

> please refer to "Device Model Parameters".

Author: ok

``probe_modules``
    Install necessary modules before launching a Post-launched VM. For
    example, pci_stub is used to provide a stub pci driver which does nothing
    on attached pci devices. and passthrough
    PCIe device will be unbind from original driver and bind to the stub, so

>     that it can be safely controlled by Post-launched VM. TODO

Reviewer 2: Why the `TODO` in the end?

Author: will drop it.

``offline_cpus <cpu_apicid>...``
    This is called if we are launching an RTVM or VM whose scheduler is
    "SCHED_NOOP". In both situations, CPU sharing between multiple VMs are
    prevented.
    This function will trigger linux cpu offline first, and then inform
    hypervisor through HSM. Hypervisor will offline VCPU and freeze VCPU
    thread.

``unbind_device <bdf>``
    Unbind a PCIe device with specified BDF (bus, device and function) number
    from its original driver and re-bind it to the pci-stub driver. After that
    linux will not operate on that device any more and it can be passthrough
    to User VM safely.

``create_tap <tap>``
    Create or reuse the tap interface which is attached to acrn-br0 bridge.
    acrn-br0 is registered to systemd-networkd.service after installing acrn
    deb package. This is used to add ``virtio-net`` interface to User VM.
    ``virtio-net`` interfaces for all User VMs are virtually connected to a
    subnet behind the acrn bridge.

``mount_partition <partition>``

>     Mount specified partition to a temporary directory created by ``mktemp -d``,

Author: Configurator allows user to give a rootfs file inside an image file as virtio-blk backend. The typical usage is assigning the "/boot/initrd.img*" inside a raw image file to User VM. The format of the string in configurator virtio-blk input box is "<imgfile>:<path/to/rootfs>". The generated launch script will mount the <imgfile> to a temporary directory first so that acrn-dm could access the inner rootfs. After the acrn-dm instance exits, unmount_partition will be called.

May I add this explanation ?

    and return the temporary directory for later unmount.

``unmount_partition <dir>``
    Unmount partition from specified directory.

``add_cpus <cpu_apicid>...``
    Return acrn-dm parameter fregment to set ``cpu_affinity``. The argument is

>     an array of the CPU lapic id, not the logical id in ServiceVM.

Reviewer 2: Better talk about the "processor number" vs. logical ID?

I would also refer back to the `cpu_affinity` description above for more details, it's all in there already,

>     an array of the CPU lapic id, not the logical id in ServiceVM.

Author: Yes, just refer back is ok

   * - ``wdt-i6300esb``
     - Emulated i6300ESB PCI Watch Dog Timer (WDT) Intel processors use to
       monitor User VMs.

Launch Script
*************

Launch script is used to launch a User VM from the ServiceVM command line. It
is generated by configurator according to several settings for a User VM. With

> launch script, user does not bother tweaking acrn-dm parameters.

Reviewer: Should we map the ACRN Configurator settings to the launch script / DM options? Would that be useful to users? Thoughts @gvancuts @dbkinder?

Reviewer 3: > Should we map the ACRN Configurator settings to the launch script / DM options? Would that be useful to users? Thoughts @gvancuts @dbkinder?

Yes, but we could update with this information after the 3.0 release (I think).  For example, we can note that the ``-m, --memsize <memory_size>`` parameter gets its value from the post-launched VM "Memory Size" basic option. (Or vice versa, we include in the "Memory Size" option description that this sets the -memsize acrn-dm parameter).  Or both. More discussion.

Reviewer: Agree that it can be a future enhancement.

``add_cpus <cpu_apicid>...``
    Return acrn-dm parameter fregment to set ``cpu_affinity``. The argument is
    an array of the CPU lapic id, not the logical id in ServiceVM.
    ``offline_cpus`` is called if the User VM is an RTVM or its scheduler is
    ``SCHED_NOOP``.

``add_interrupt_storm_monitor <threshold_per_sec> <probe_period_in_sec> <inject_delay_in_ms> <delay_duration_in_ms>``
    This is added if PCIe devices except integrated GPU are passthrough to the
    User VM to monitor if interrupt storm occurred on those devices.
    This function and parameters are not visible in configurator and handled
    by config scripts.
    It returns acrn-dm command line segment to set ``intr_monitor``.

``add_logger_settings console=<n> kmsg=<n> disk=<n>``
    set log level of each logging channel. (console, kmsg, disk). This is not
    exposed to user in configurator.

``add_virtual_device <slot> <kind> <options>``
    Add specified kind of virtual device to the specified PCIe device slot.
    Some devices need options to configure further behaviors. <slot> numbers
    for virtual devices and passthrough devices are automatically allocated

>     by `launch_cfg_gen.py`.

Reviewer 2: ```suggestion
    by ``launch_cfg_gen.py``.
```

    Typical use cases:

    - ``hostbridge``
        PCIe host bridge. <slot> must be 0.

    - ``uart vuart_idx:<int>``
        Add a PCIe vuart with specified index.

    - ``xhci <bus>-<port>[:<bus>-<port>]...``

>         Config usb mediator. A list of usb devices each specified by

Reviewer 2: ```suggestion
        Config USB mediator. A list of USB ports each specified by
```

        ``<bus>-<port>`` will be connected to the User VM.

>     - ``virtio-net tap=<tapname>[,vhost],mac_seed=<str>``

Reviewer 2: Does it also take `mac` instead of `mac_seed`?

>         The tap should already be created by ``create_tap``.

Reviewer 2: ```suggestion
        The TAP should already be created by ``create_tap``.
```

    - ``virtio-blk <imgfile>[,writethru|writeback|ro]``

>         Add an virtio block device to User VM. The backend is an raw image

Reviewer 2: ```suggestion
        Add a virtio block device to User VM. The backend is a raw image
```

VM. With launch script, user does not bother tweaking ``acrn-dm`` command line
parameters.

Most settings for User VMs are used at launch time. They are first stored into
``scenario.xml`` and then parsed by
``misc/config_tools/launch_config/launch_cfg_gen.py``
to add shell commands into launch script according to the template
``misc/config_tools/launch_config/launch_script_template.sh``.
The template defines following helper functions to do system settings or to
generate an ``acrn-dm`` command line parameter. For details about all
``acrn-dm`` parameters, please refer to :ref:`acrn-dm_parameters`.

``probe_modules``
    Install necessary modules before launching a Post-launched VM. For
    example, ``pci_stub`` is used to provide a stub pci driver which does
    nothing on attached pci devices. Passthrough PCIe devices will be unbound
    from original driver and bound to the stub, so that they can be safely
    controlled by the User VM.

``offline_cpus <cpu_apicid>...``
    This is called if we are launching an RTVM or VM whose scheduler is
    "SCHED_NOOP". In both situations, CPU sharing between multiple VMs are
    prevented.

>     This function will trigger linux cpu offline first, and then inform

Author: I mean cpu offline action done by Serivce VM kernel.

    hypervisor through Hypervisor Service Module (HSM). The hypervisor will
    offline vCPU and freeze vCPU thread.

``unbind_device <bdf>``
    Unbind a PCIe device with specified BDF (bus, device and function) number
    from its original driver and re-bind it to the pci-stub driver. After that
    the Service VM kernel will not operate on that device any more and it can
    be passed through to User VM safely.

``create_tap <tap>``
    Create or reuse the tap interface which is attached to ``acrn-br0`` bridge.

>     ``acrn-br0`` is registered to ``systemd-networkd.service`` after installing

Author: ACRN Debian package just registers the bridge to "systemd-networkd" service. But we have to make sure the systemd service is enabled. How about adding: "User needs to enable the service to create the bridge".

    ACRN Debian package (``.deb``). This is used to add ``virtio-net``
    interface to User VM. ``virtio-net`` interfaces for all User VMs are
    virtually connected to a subnet behind the ACRN bridge.

``mount_partition <partition>``
    Mount specified partition to a temporary directory created by ``mktemp -d``,
    and return the temporary directory for later unmount.
    Typically this function is called to mount an image file in order to use
    inner rootfs file as ``virtio-blk`` backend. For example, user could set
    "<imgfile>:/boot/initrd.img*" in the ``virtio-blk`` input box in the ACRN
    Configurator. After the ``acrn-dm`` instance exits, ``unmount_partition``
    will be called to unmount the image file.

``unmount_partition <dir>``
    Unmount partition from specified directory.

``add_cpus <cpu_apicid>...``
    Return an ``acrn-dm`` command line parameter fragment to set

>     ``cpu_affinity``. Please refer to `cpu_affinity`_ for details.

Reviewer 3: ```suggestion
    ``cpu_affinity``. Refer to `cpu_affinity`_ for details.
```

    ``offline_cpus`` is called if the User VM is an RTVM or its scheduler is
    ``SCHED_NOOP``.

``add_interrupt_storm_monitor <threshold_per_sec> <probe_period_in_sec> <inject_delay_in_ms> <delay_duration_in_ms>``
    This is added if PCIe devices except integrated GPU are passed through to
    the User VM to monitor if interrupt storm occurred on those devices.
    This function and parameters are not visible in the ACRN Configurator and
    handled by config scripts.
    It returns acrn-dm command line segment to set ``intr_monitor``.

``add_logger_settings console=<n> kmsg=<n> disk=<n>``

>     set log level of each logging channel. (console, kmsg, disk). This is not

Author: "ACRN log level, Intel Trace Hub log level, and Serial console log level" are for acrn hypervisor. "add_logger_settings" sets log levels of the acrn-dm instance.

    exposed to user in the ACRN Configurator.

``add_virtual_device <slot> <kind> <options>``
    Add specified kind of virtual device to the specified PCIe device slot.
    Some devices need options to configure further behaviors. <slot> numbers
    for virtual devices and passthrough devices are automatically allocated
    by ``launch_cfg_gen.py``.

    Typical use cases:

    - ``hostbridge``
        PCIe host bridge. <slot> must be 0.

    - ``uart vuart_idx:<int>``
        Add a PCIe vuart with specified index.

    - ``xhci <bus>-<port>[:<bus>-<port>]...``
        Config USB mediator. A list of USB ports each specified by
        ``<bus>-<port>`` will be connected to the User VM.

    - ``virtio-net tap=<tapname>[,vhost],mac_seed=<str>``
        The TAP should already be created by ``create_tap``.

    - ``virtio-blk <imgfile>[,writethru|writeback|ro]``
        Add a virtio block device to User VM. The backend is a raw image
        file. Options can be specified to control access right.

>     For all types of virtual devices and options, please refer to

Reviewer 3: ```suggestion
    For all types of virtual devices and options, refer to
```

    :ref:`emul_config`.

``add_passthrough_device <slot> <bus>/<device>/<function> <options>``
    Passthrough PCIe device to User VM in specified ``<slot>``.
    Some kinds of devices may need extra ``<options>`` to control internal
    behavior. Please refer to ``passthru`` section in :ref:`emul_config`.

These functions in template are copied to target launch script. Then
``launch_cfg_gen.py`` generates following dynamic part. It first defines
necessary variables like ``vm_type`` and ``scheduler``. Then above functions
are called to construct the ``dm_params`` per user settings in ``scenario.xml``.
And last, ``acrn-dm`` is executed to launch a User VM with these parameters.

> ******************

Reviewer 3: There's not enough information in this section to include it now, including a usage model for "Dynamic VMs" and a real example with enough details.  I'd recommend this section be deleted and we look into these "Dynamic VMs" after the 3.0 release.  @gvancuts @NanlinXie Your thoughts?

Author: Yes, I agree. Further discussions on Dynamic VMs are required.

    Typically this function is called to mount an image file in order to use
    inner rootfs file as ``virtio-blk`` backend. For example, user could set
    "<imgfile>:/boot/initrd.img*" in the ``virtio-blk`` input box in the ACRN
    Configurator. After the ``acrn-dm`` instance exits, ``unmount_partition``
    will be called to unmount the image file.

``unmount_partition <dir>``
    Unmount partition from specified directory.

``add_cpus <cpu_apicid>...``
    Return an ``acrn-dm`` command line parameter fragment to set
    ``cpu_affinity``. Refer to `cpu_affinity`_ for details.
    ``offline_cpus`` is called if the User VM is an RTVM or its scheduler is
    ``SCHED_NOOP``.

``add_interrupt_storm_monitor <threshold_per_sec> <probe_period_in_sec> <inject_delay_in_ms> <delay_duration_in_ms>``
    This is added if PCIe devices, other than an integrated GPU, are passed through to
    the User VM to monitor if interrupt storm occurred on those devices.
    This function and parameters are not visible in the ACRN Configurator and
    handled by config scripts.
    It returns acrn-dm command line segment to set ``intr_monitor``.

``add_logger_settings console=<n> kmsg=<n> disk=<n>``

>     set log level of each logging channel. (console, kmsg, disk). This is not

Reviewer 3: ```suggestion
    set log level of each ``acrn-dm`` logging channel: console, kmsg, and disk. These settings are not
```

    exposed to user in the ACRN Configurator.

``add_virtual_device <slot> <kind> <options>``
    Add specified kind of virtual device to the specified PCIe device slot.

>     Some devices need options to configure further behaviors. <slot> numbers

Reviewer 3: ```suggestion
    Some devices need options to configure further behaviors. ``<slot>`` numbers
```

    for virtual devices and passthrough devices are automatically allocated
    by ``launch_cfg_gen.py``.

    Typical use cases:

    - ``hostbridge``

>         PCIe host bridge. <slot> must be 0.

Reviewer 3: ```suggestion
        PCIe host bridge. ``<slot>`` must be 0.
```
